File: test/specs/test_host.zig
/// Test host implementation for execution-specs tests
/// This maintains multi-account state (balances, code, storage) for testing
const std = @import("std");
const primitives = @import("primitives");
const evm_mod = @import("evm");
const Address = primitives.Address.Address;
const HostInterface = evm_mod.HostInterface;
const CallResult = evm_mod.CallResult;

/// Storage slot key for tracking
pub const StorageSlotKey = struct {
    address: Address,
    slot: u256,

    pub fn hash(self: @This(), key: StorageSlotKey) u32 {
        _ = self;
        var hasher = std.hash.Wyhash.init(0);
        hasher.update(&key.address.bytes);
        hasher.update(std.mem.asBytes(&key.slot));
        return @truncate(hasher.final());
    }

    pub fn eql(self: @This(), a: StorageSlotKey, b: StorageSlotKey, b_index: usize) bool {
        _ = self;
        _ = b_index;
        return a.address.equals(b.address) and a.slot == b.slot;
    }
};

const AddressContext = std.array_hash_map.AutoContext(Address);
const StorageContext = std.hash_map.AutoContext(StorageSlotKey);

pub const TestHost = struct {
    const Self = @This();

    allocator: std.mem.Allocator,
    balances: std.AutoHashMap(Address, u256),
    code: std.AutoHashMap(Address, []const u8),
    storage: std.AutoHashMap(StorageSlotKey, u256),
    nonces: std.AutoHashMap(Address, u64),

    pub fn init(allocator: std.mem.Allocator) !Self {
        return .{
            .allocator = allocator,
            .balances = std.AutoHashMap(Address, u256).init(allocator),
            .code = std.AutoHashMap(Address, []const u8).init(allocator),
            .storage = std.AutoHashMap(StorageSlotKey, u256).init(allocator),
            .nonces = std.AutoHashMap(Address, u64).init(allocator),
        };
    }

    pub fn deinit(self: *Self) void {
        // Free all allocated code
        var it = self.code.valueIterator();
        while (it.next()) |code_ptr| {
            self.allocator.free(code_ptr.*);
        }
        self.balances.deinit();
        self.code.deinit();
        self.storage.deinit();
        self.nonces.deinit();
    }

    pub fn hostInterface(self: *Self) HostInterface {
        return .{
            .ptr = self,
            .vtable = &.{
                .innerCall = innerCall,
                .getBalance = getBalance,
                .setBalance = setBalanceVTable,
                .getCode = getCode,
                .setCode = setCodeVTable,
                .getStorage = getStorage,
                .setStorage = setStorage,
                .getNonce = getNonceVTable,
                .setNonce = setNonceVTable,
            },
        };
    }

    // Pre-state setup methods
    pub fn setBalance(self: *Self, address: Address, balance: u256) !void {
        try self.balances.put(address, balance);
    }

    pub fn setCode(self: *Self, address: Address, code: []const u8) !void {
        const owned_code = try self.allocator.dupe(u8, code);
        try self.code.put(address, owned_code);
    }

    pub fn setStorageSlot(self: *Self, address: Address, slot: u256, value: u256) !void {
        const key = StorageSlotKey{ .address = address, .slot = slot };
        // EVM spec: storage slots with value 0 should be deleted, not stored
        if (value == 0) {
            _ = self.storage.remove(key);
        } else {
            try self.storage.put(key, value);
        }
    }

    pub fn setNonce(self: *Self, address: Address, nonce: u64) !void {
        try self.nonces.put(address, nonce);
    }

    // Post-state validation methods
    pub fn getNonce(self: *Self, address: Address) u64 {
        return self.nonces.get(address) orelse 0;
    }

    // HostInterface vtable implementations
    fn innerCall(ptr: *anyopaque, gas: u64, address: Address, value: u256, input: []const u8, call_type: HostInterface.CallType) CallResult {
        _ = input;
        _ = call_type;
        const self: *Self = @ptrCast(@alignCast(ptr));

        // Simple stub implementation:
        // 1. Don't execute code (would need full EVM for proper nested calls)
        // 2. For now, just return success without executing
        // TODO: Implement proper nested call execution with code execution

        // Note: Balance transfers should be handled by the calling EVM/frame,
        // not here in innerCall. The caller should deduct from sender and add to receiver.
        // For now, we just return success to allow tests to proceed.
        _ = self;
        _ = address;
        _ = value;

        return .{
            .success = true,
            .gas_left = gas,
            .output = &[_]u8{},
        };
    }

    fn getBalance(ptr: *anyopaque, address: Address) u256 {
        const self: *Self = @ptrCast(@alignCast(ptr));
        return self.balances.get(address) orelse 0;
    }

    fn setBalanceVTable(ptr: *anyopaque, address: Address, balance: u256) void {
        const self: *Self = @ptrCast(@alignCast(ptr));
        self.balances.put(address, balance) catch {
            return;
        };
    }

    fn getCode(ptr: *anyopaque, address: Address) []const u8 {
        const self: *Self = @ptrCast(@alignCast(ptr));
        return self.code.get(address) orelse &[_]u8{};
    }

    fn setCodeVTable(ptr: *anyopaque, address: Address, code: []const u8) void {
        const self: *Self = @ptrCast(@alignCast(ptr));
        // Note: We need to duplicate code to own it, as the passed slice may be temporary
        // For empty code, just clear the entry
        if (code.len == 0) {
            if (self.code.fetchRemove(address)) |kv| {
                self.allocator.free(kv.value);
            }
            return;
        }
        // Free existing code if present
        if (self.code.fetchRemove(address)) |kv| {
            self.allocator.free(kv.value);
        }
        // Store new code (duplicate to own the memory - the caller may free their copy)
        const owned_code = self.allocator.dupe(u8, code) catch {
            return; // In a test context, we should not fail silently, but the interface doesn't allow errors
        };
        self.code.put(address, owned_code) catch {
            self.allocator.free(owned_code);
            return;
        };
    }

    fn getStorage(ptr: *anyopaque, address: Address, slot: u256) u256 {
        const self: *Self = @ptrCast(@alignCast(ptr));
        const key = StorageSlotKey{ .address = address, .slot = slot };
        return self.storage.get(key) orelse 0;
    }

    fn setStorage(ptr: *anyopaque, address: Address, slot: u256, value: u256) void {
        const self: *Self = @ptrCast(@alignCast(ptr));
        const key = StorageSlotKey{ .address = address, .slot = slot };
        // EVM spec: storage slots with value 0 should be deleted, not stored
        if (value == 0) {
            _ = self.storage.remove(key);
        } else {
            self.storage.put(key, value) catch {
                // In a test context, we should not fail silently
                // But the interface doesn't allow errors
                return;
            };
        }
    }

    fn getNonceVTable(ptr: *anyopaque, address: Address) u64 {
        const self: *Self = @ptrCast(@alignCast(ptr));
        return self.getNonce(address);
    }

    fn setNonceVTable(ptr: *anyopaque, address: Address, nonce: u64) void {
        const self: *Self = @ptrCast(@alignCast(ptr));
        self.setNonce(address, nonce) catch {
            // In a test context, we should not fail silently
            // But the interface doesn't allow errors
            return;
        };
    }
};

File: test/specs/test-host.ts
/**
 * Test Host Implementation
 *
 * In-memory state backend for EVM testing, implementing the HostInterface.
 * Provides storage for balances, nonces, code, storage, and logs.
 *
 * Port of test/specs/test_host.zig
 */

import type { HostInterface, Address } from '../../src/host';
import type { Log } from '../../src/call-result';

/**
 * TestHost: In-memory EVM state backend for testing
 *
 * Implements HostInterface with simple Map-based storage.
 * All state is stored in memory and cleared between tests.
 */
export class TestHost implements HostInterface {
  private balances: Map<string, bigint> = new Map();
  private nonces: Map<string, bigint> = new Map();
  private code: Map<string, Uint8Array> = new Map();
  private storage: Map<string, Map<string, bigint>> = new Map();
  private logs: Log[] = [];

  /**
   * Convert address to string key for Map storage
   */
  private addressKey(address: Address): string {
    return Array.from(address)
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
  }

  /**
   * Convert slot (bigint) to string key
   */
  private slotKey(slot: bigint): string {
    return slot.toString(16).padStart(64, '0');
  }

  /**
   * Get balance for an address
   */
  getBalance(address: Address): bigint {
    const key = this.addressKey(address);
    return this.balances.get(key) ?? 0n;
  }

  /**
   * Set balance for an address
   */
  setBalance(address: Address, balance: bigint): void {
    const key = this.addressKey(address);
    if (balance === 0n) {
      this.balances.delete(key);
    } else {
      this.balances.set(key, balance);
    }
  }

  /**
   * Get nonce for an address
   */
  getNonce(address: Address): bigint {
    const key = this.addressKey(address);
    return this.nonces.get(key) ?? 0n;
  }

  /**
   * Set nonce for an address
   */
  setNonce(address: Address, nonce: number | bigint): void {
    const key = this.addressKey(address);
    const nonceValue = typeof nonce === 'number' ? BigInt(nonce) : nonce;
    if (nonceValue === 0n) {
      this.nonces.delete(key);
    } else {
      this.nonces.set(key, nonceValue);
    }
  }

  /**
   * Get code for an address
   */
  getCode(address: Address): Uint8Array {
    const key = this.addressKey(address);
    return this.code.get(key) ?? new Uint8Array(0);
  }

  /**
   * Set code for an address
   */
  setCode(address: Address, code: Uint8Array): void {
    const key = this.addressKey(address);
    if (code.length === 0) {
      this.code.delete(key);
    } else {
      // Store a copy to avoid external modifications
      this.code.set(key, new Uint8Array(code));
    }
  }

  /**
   * Get storage value for an address and slot
   */
  getStorage(address: Address, slot: bigint): bigint {
    const addrKey = this.addressKey(address);
    const slotMap = this.storage.get(addrKey);
    if (!slotMap) {
      return 0n;
    }
    const slotKey = this.slotKey(slot);
    return slotMap.get(slotKey) ?? 0n;
  }

  /**
   * Set storage value for an address and slot
   */
  setStorage(address: Address, slot: bigint, value: bigint): void {
    const addrKey = this.addressKey(address);
    let slotMap = this.storage.get(addrKey);

    if (!slotMap) {
      if (value === 0n) {
        // No need to create a map for a zero value
        return;
      }
      slotMap = new Map();
      this.storage.set(addrKey, slotMap);
    }

    const slotKey = this.slotKey(slot);
    if (value === 0n) {
      slotMap.delete(slotKey);
      // Clean up empty maps
      if (slotMap.size === 0) {
        this.storage.delete(addrKey);
      }
    } else {
      slotMap.set(slotKey, value);
    }
  }

  /**
   * Add a log entry
   */
  addLog(log: Log): void {
    this.logs.push({
      address: new Uint8Array(log.address),
      topics: [...log.topics],
      data: new Uint8Array(log.data),
    });
  }

  /**
   * Get all logs
   */
  getLogs(): Log[] {
    return this.logs;
  }

  /**
   * Clear all logs
   */
  clearLogs(): void {
    this.logs = [];
  }

  /**
   * Clear all state (for test cleanup)
   */
  clear(): void {
    this.balances.clear();
    this.nonces.clear();
    this.code.clear();
    this.storage.clear();
    this.logs = [];
  }

  /**
   * Get all accounts with balances (for debugging/testing)
   */
  getAllAccounts(): Array<{ address: string; balance: bigint }> {
    const accounts: Array<{ address: string; balance: bigint }> = [];
    for (const [addrKey, balance] of this.balances.entries()) {
      accounts.push({ address: `0x${addrKey}`, balance });
    }
    return accounts;
  }

  /**
   * Get all storage for an address (for debugging/testing)
   */
  getAllStorage(address: Address): Array<{ slot: bigint; value: bigint }> {
    const addrKey = this.addressKey(address);
    const slotMap = this.storage.get(addrKey);
    if (!slotMap) {
      return [];
    }

    const entries: Array<{ slot: bigint; value: bigint }> = [];
    for (const [slotKey, value] of slotMap.entries()) {
      entries.push({ slot: BigInt(`0x${slotKey}`), value });
    }
    return entries;
  }
}
